compiler/backend/src/org/jetbrains/kotlin/codegen/inline/InlineCodegen.kt

```diff
@@ -433,145 +426,64 @@ abstract class InlineCodegen<out T : BaseExpressionCodegen>(
                 findVirtualFile(state, containerId)?.contentsToByteArray()
                     ?: throw IllegalStateException("Couldn't find declaration file for $containerId")
             }
-
-            val classType = AsmUtil.asmTypeByClassId(containerId)
-            val methodNode = getMethodNode(bytes, asmMethod.name, asmMethod.descriptor, classType)
-            if (methodNode == null && requiresFunctionNameManglingForReturnType(callableDescriptor)) {
-                val nameWithoutManglingSuffix = asmMethod.name.stripManglingSuffixOrNull()
-                if (nameWithoutManglingSuffix != null) {
-                    val methodWithoutMangling = getMethodNode(bytes, nameWithoutManglingSuffix, asmMethod.descriptor, classType)
-                    if (methodWithoutMangling != null) return methodWithoutMangling
-                }
-                return getMethodNode(bytes, "$nameWithoutManglingSuffix-impl", asmMethod.descriptor, classType)
+            val resultInCache = state.inlineCache.methodNodeById.getOrPut(MethodId(containerType.descriptor, asmMethod)) {
+                getMethodNode(containerType, bytes, asmMethod.name, asmMethod.descriptor, isSuspend, isMangled)
             }
-
-            return methodNode
-        }
-
-        private fun String.stripManglingSuffixOrNull(): String? {
-            val dashIndex = indexOf('-')
-            return if (dashIndex < 0) null else substring(0, dashIndex)
+            return SMAPAndMethodNode(cloneMethodNode(resultInCache.node), resultInCache.classSMAP)
         }
 
-        private fun isBuiltInArrayIntrinsic(callableDescriptor: CallableMemberDescriptor): Boolean {
-            if (callableDescriptor is FictitiousArrayConstructor) return true
-            val name = callableDescriptor.name.asString()
-            return (name == "arrayOf" || name == "emptyArray") && callableDescriptor.containingDeclaration.let { container ->
-                container is PackageFragmentDescriptor && container.fqName == StandardNames.BUILT_INS_PACKAGE_FQ_NAME
+        private fun getMethodNode(
+            owner: Type,
+            bytes: ByteArray,
+            name: String,
+            descriptor: String,
+            isSuspend: Boolean,
+            isMangled: Boolean
+        ): SMAPAndMethodNode {
+            getMethodNode(owner, bytes, name, descriptor, isSuspend)?.let { return it }
+            if (isMangled) {
+                // Compatibility with old inline class ABI versions.
+                val dashIndex = name.indexOf('-')
+                val nameWithoutManglingSuffix = if (dashIndex > 0) name.substring(0, dashIndex) else name
+                if (nameWithoutManglingSuffix != name) {
+                    getMethodNode(owner, bytes, nameWithoutManglingSuffix, descriptor, isSuspend)?.let { return it }
+                }
+                getMethodNode(owner, bytes, "$nameWithoutManglingSuffix-impl", descriptor, isSuspend)?.let { return it }
             }
```

Comment:
Neither the old code nor the refactored version is 100% correct, since suffixes such as `$default` and `$$forInline` go after the hash. This commit fixes the `$$forInline` case, but something like `x-abcdef$default` is still transformed into `x-impl` by this code. I suppose it doesn't matter all that much since nobody should've really been using this ABI anyway.

Reply:
@pyos  Could you add test for fixed `$$forInline` case? And maybe muted one for `x-abcdef$default` ?

Reply:
this requires compiling one module with feature `MangleClassMembersReturningInlineClasses` disabled and the other with it enabled, does the test infrastructure allow that?

Reply:
actually this doesn't fix `$$forInline` either, because for suspend functions `requiresFunctionNameManglingForReturnType` always return false since it checks the return type of the suspend view, which is always `Any?`

Reply:
is there even any point in having this branch anymore if the feature was enabled by default from 1.4? are language features ever removed at all? there are some dating back to 1.1...

Reply:
> does the test infrastructure allow that?
There is example of different `-Xjvm-default` options for different modules
https://github.com/JetBrains/kotlin/blob/5075484c8e52fe73b4f29d862c3c5730f9b9d149/compiler/testData/codegen/box/compileKotlinAgainstKotlin/jvm8/defaults/interop/likeMemberClash.kt



>is there even any point in having this branch anymore if the feature was enabled by default from 1.4?

Compiler until 1.6 would support 1.3 target. This branch couldn't be removed now. But test in this cases actually is not necessery

> are language features ever removed at all? there are some dating back to 1.1...
Yeah...nobody take care of them, but yes they could be removed